- Coger subsets aleatorios
- Ver si usar item based collab porque tiene mas escalabilidad, user based necesita calcular la similitud de todos los pares de usuarios

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [2]:
#df_users = pd.read_csv('data/train_triplets.txt', sep='\t', header=None, names=['user_id', 'song_id', 'play_count'])
df_users = pd.read_csv('data/Million Song Dataset kaggle/User Listening History.csv')

In [3]:
df_music = pd.read_csv('data/Million Song Dataset kaggle/Music Info.csv')

In [4]:
df_users

track_id                                   user_id  \
0        TRIRLYL128F42539D1  b80344d063b5ccb3212f76538f3d9e43d87dca9e   
1        TRFUPBA128F934F7E1  b80344d063b5ccb3212f76538f3d9e43d87dca9e   
2        TRLQPQJ128F42AA94F  b80344d063b5ccb3212f76538f3d9e43d87dca9e   
3        TRTUCUY128F92E1D24  b80344d063b5ccb3212f76538f3d9e43d87dca9e   
4        TRHDDQG12903CB53EE  b80344d063b5ccb3212f76538f3d9e43d87dca9e   
...                     ...                                       ...   
9711296  TRBKFKL128E078ED76  b7815dbb206eb2831ce0fe040d0aa537e2e800f7   
9711297  TRYFXPG128E078ECBD  b7815dbb206eb2831ce0fe040d0aa537e2e800f7   
9711298  TROBUUZ128F4263002  b7815dbb206eb2831ce0fe040d0aa537e2e800f7   
9711299  TROEWXC128F148C83E  b7815dbb206eb2831ce0fe040d0aa537e2e800f7   
9711300  TROHQNO128F146ECCB  b7815dbb206eb2831ce0fe040d0aa537e2e800f7   

         playcount  
0                1  
1                1  
2                1  
3                1  
4                1  
...            ...  
9711296          1  
9711297          1  
9711298          1  
9711299          1  
9711300          1  

[9711301 rows x 3 columns]

In [5]:
df_users.dtypes

track_id     object
user_id      object
playcount     int64
dtype: object

In [6]:
num_distinct_users = df_users['user_id'].nunique()
num_distinct_users

962037

In [7]:
data_df = df_users.groupby('track_id').size().reset_index(name='count')

In [8]:
data_df

track_id  count
0      TRAAAED128E0783FAB    161
1      TRAAAHD128F42635A5      2
2      TRAAAQO12903CD8E1C      1
3      TRAABJS128F9325C99    945
4      TRAABXA128F9326351     18
...                   ...    ...
30454  TRZZYMU128E0792400   1140
30455  TRZZZCL128F428BB80     19
30456  TRZZZCN128F9317A03     15
30457  TRZZZHL128F9329CFB    116
30458  TRZZZRJ128F42819AF    857

[30459 rows x 2 columns]

In [9]:
df_music.head()

track_id             name           artist  \
0  TRIOREW128F424EAF0   Mr. Brightside      The Killers   
1  TRRIVDJ128F429B0E8       Wonderwall            Oasis   
2  TROUVHL128F426C441  Come as You Are          Nirvana   
3  TRUEIND128F93038C4      Take Me Out  Franz Ferdinand   
4  TRLNZBD128F935E4D8            Creep        Radiohead   

                                 spotify_preview_url              spotify_id  \
0  https://p.scdn.co/mp3-preview/4d26180e6961fd46...  09ZQ5TmUG8TSL56n0knqrj   
1  https://p.scdn.co/mp3-preview/d012e536916c927b...  06UfBBDISthj1ZJAtX4xjj   
2  https://p.scdn.co/mp3-preview/a1c11bb1cb231031...  0keNu0t0tqsWtExGM3nT1D   
3  https://p.scdn.co/mp3-preview/399c401370438be4...  0ancVQ9wEcHVd0RrGICTE4   
4  https://p.scdn.co/mp3-preview/e7eb60e9466bc3a2...  01QoK9DA7VTeTSE3MNzp4I   

                                                tags genre  year  duration_ms  \
0  rock, alternative, indie, alternative_rock, in...   NaN  2004       222200   
1  rock, alternative, indie, pop, alternative_roc...   NaN  2006       258613   
2   rock, alternative, alternative_rock, 90s, grunge   RnB  1991       218920   
3  rock, alternative, indie, alternative_rock, in...   NaN  2004       237026   
4  rock, alternative, indie, alternative_rock, in...   RnB  2008       238640   

   danceability  ...  key  loudness  mode  speechiness  acousticness  \
0         0.355  ...    1    -4.360     1       0.0746      0.001190   
1         0.409  ...    2    -4.373     1       0.0336      0.000807   
2         0.508  ...    4    -5.783     0       0.0400      0.000175   
3         0.279  ...    9    -8.851     1       0.0371      0.000389   
4         0.515  ...    7    -9.935     1       0.0369      0.010200   

   instrumentalness  liveness  valence    tempo  time_signature  
0          0.000000    0.0971    0.240  148.114               4  
1          0.000000    0.2070    0.651  174.426               4  
2          0.000459    0.0878    0.543  120.012               4  
3          0.000655    0.1330    0.490  104.560               4  
4          0.000141    0.1290    0.104   91.841               4  

[5 rows x 21 columns]

In [10]:
df_music_info = df_music[['track_id', 'name', 'artist']]
df_music_info.head()

track_id             name           artist
0  TRIOREW128F424EAF0   Mr. Brightside      The Killers
1  TRRIVDJ128F429B0E8       Wonderwall            Oasis
2  TROUVHL128F426C441  Come as You Are          Nirvana
3  TRUEIND128F93038C4      Take Me Out  Franz Ferdinand
4  TRLNZBD128F935E4D8            Creep        Radiohead

In [11]:
df_music_count = df_music_info.merge(
    df_users.groupby('track_id').size().reset_index(name='playcount'),
    on='track_id',
    how='left'
)
df_music_count = df_music_count.fillna(0)
print(df_music_count.shape)
df_music_count.head()

(50683, 4)


track_id             name           artist  playcount
0  TRIOREW128F424EAF0   Mr. Brightside      The Killers       41.0
1  TRRIVDJ128F429B0E8       Wonderwall            Oasis      378.0
2  TROUVHL128F426C441  Come as You Are          Nirvana        0.0
3  TRUEIND128F93038C4      Take Me Out  Franz Ferdinand       20.0
4  TRLNZBD128F935E4D8            Creep        Radiohead        0.0

In [12]:
# df_music_count = df_music_info.merge(
#     df_users.groupby('track_id').size().reset_index(name='playcount'),
#     on='track_id',
#     how='inner'
# )
# print(df_music_count.shape)
# df_music_count.head()

In [117]:
df_users_agg = df_users.groupby('user_id')['playcount'].agg(
    #total_playcount='sum',
    max_playcount='max'
).reset_index()
df_users_agg = df_users_agg.rename(columns={'playcount': 'max_playcount'})

In [118]:
df_users_rating = df_users.merge(df_users_agg, on='user_id')

In [119]:
df_users_rating['rating'] = df_users_rating['playcount'] / df_users_rating['max_playcount']
df_users_rating[:50]

track_id                                   user_id  playcount  \
0   TRIRLYL128F42539D1  b80344d063b5ccb3212f76538f3d9e43d87dca9e          1   
1   TRFUPBA128F934F7E1  b80344d063b5ccb3212f76538f3d9e43d87dca9e          1   
2   TRLQPQJ128F42AA94F  b80344d063b5ccb3212f76538f3d9e43d87dca9e          1   
3   TRTUCUY128F92E1D24  b80344d063b5ccb3212f76538f3d9e43d87dca9e          1   
4   TRHDDQG12903CB53EE  b80344d063b5ccb3212f76538f3d9e43d87dca9e          1   
5   TRGRDEC128F423C07D  b80344d063b5ccb3212f76538f3d9e43d87dca9e          1   
6   TRRYCBO128F932A2C7  b80344d063b5ccb3212f76538f3d9e43d87dca9e          1   
7   TRUTULC128F4293712  b80344d063b5ccb3212f76538f3d9e43d87dca9e          1   
8   TRAAHSY128F147BB5C  b80344d063b5ccb3212f76538f3d9e43d87dca9e          2   
9   TRDLMWP128F426BF6C  b80344d063b5ccb3212f76538f3d9e43d87dca9e          1   
10  TRMSZXT128F92F7816  b80344d063b5ccb3212f76538f3d9e43d87dca9e          1   
11  TRVODUZ128F934D094  b80344d063b5ccb3212f76538f3d9e43d87dca9e          1   
12  TRLXSNR128F429361D  b80344d063b5ccb3212f76538f3d9e43d87dca9e          1   
13  TRPUGUW128F426BF6F  b80344d063b5ccb3212f76538f3d9e43d87dca9e          1   
14  TRAUCNU128F42671EB  b80344d063b5ccb3212f76538f3d9e43d87dca9e          1   
15  TRVVWSW128F4292931  b80344d063b5ccb3212f76538f3d9e43d87dca9e          1   
16  TRADVZX128F426BF79  b80344d063b5ccb3212f76538f3d9e43d87dca9e          1   
17  TRDCZUX128F4263A74  b80344d063b5ccb3212f76538f3d9e43d87dca9e          1   
18  TRUYKTB128F1459197  b80344d063b5ccb3212f76538f3d9e43d87dca9e          1   
19  TRQLOHS128F42435AE  b80344d063b5ccb3212f76538f3d9e43d87dca9e          1   
20  TRCTFBL128F42934F3  85c1f87fea955d09b4bec2e36aee110927aedf9a          2   
21  TRPGYLT128F428AD02  85c1f87fea955d09b4bec2e36aee110927aedf9a          3   
22  TRAEGXU128F9304D31  bd4c6e843f00bd476847fb75c47b4fb430a06856          1   
23  TRFNNFP128F146CB4B  bd4c6e843f00bd476847fb75c47b4fb430a06856          1   
24  TRWCEKE128F93191BE  bd4c6e843f00bd476847fb75c47b4fb430a06856          2   
25  TRADEOV128F9330E3B  bd4c6e843f00bd476847fb75c47b4fb430a06856          1   
26  TRTQVKS128F92EF9BE  bd4c6e843f00bd476847fb75c47b4fb430a06856          1   
27  TRSGIYX128F149F01F  969cc6fb74e076a68e36a04409cb9d3765757508          1   
28  TRXWAZC128F9314B3E  969cc6fb74e076a68e36a04409cb9d3765757508          1   
29  TRDOYFZ128F426E715  969cc6fb74e076a68e36a04409cb9d3765757508          1   
30  TRBFXMJ12903CB50F6  4bd88bfb25263a75bbdd467e74018f4ae570e5df          2   
31  TRDSFKT12903CB510F  4bd88bfb25263a75bbdd467e74018f4ae570e5df          4   
32  TRRELZC128E078ED67  4bd88bfb25263a75bbdd467e74018f4ae570e5df          6   
33  TRURIYJ12903CB50FA  4bd88bfb25263a75bbdd467e74018f4ae570e5df          3   
34  TRISTWT128F1488FBB  e006b1a48f466bf59feefed32bec6494495a4436          1   
35  TRKPWGR128E078EE06  e006b1a48f466bf59feefed32bec6494495a4436          2   
36  TRJZLOA128F930369C  e006b1a48f466bf59feefed32bec6494495a4436          4   
37  TRUZPNY128F147FD23  e006b1a48f466bf59feefed32bec6494495a4436          1   
38  TRLSADE128F9321896  e006b1a48f466bf59feefed32bec6494495a4436          1   
39  TRONYHY128F92C9D11  e006b1a48f466bf59feefed32bec6494495a4436          2   
40  TRTDKMT128F92CDD87  e006b1a48f466bf59feefed32bec6494495a4436          3   
41  TRGAOLV128E0789D40  9d6f0ead607ac2a6c2460e4d14fb439a146b7dec          2   
42  TRCIVWB128F92FDBB3  9d6f0ead607ac2a6c2460e4d14fb439a146b7dec          2   
43  TRFWGOJ128E0780C8B  9d6f0ead607ac2a6c2460e4d14fb439a146b7dec          2   
44  TRZCIWG128F4248B25  9d6f0ead607ac2a6c2460e4d14fb439a146b7dec          2   
45  TRXQJWK128F146DF76  9bb911319fbc04f01755814cb5edb21df3d1a336         16   
46  TRHNJTW128E0784786  b64cdd1a0bd907e5e00b39e345194768e330d652          1   
47  TRNJLKP128F427CE28  b64cdd1a0bd907e5e00b39e345194768e330d652          3   
48  TRZACEN128F428E742  b64cdd1a0bd907e5e00b39e345194768e330d652          1   
49  TRGASNY128F14696B0  b64cdd1a0bd907e5e00b39e345194768e330d652

In [120]:
df_users_agg[df_users_agg['user_id'] == '85c1f87fea955d09b4bec2e36aee110927aedf9a']

user_id  max_playcount
502464  85c1f87fea955d09b4bec2e36aee110927aedf9a              3

For the interaction matrix, we shall use a sparse matrix, due to users only listening to a small subset of the songs

In [122]:
from scipy.sparse import csr_matrix

We cant reproduce next line because it needs 218 GB of RAM to create the table

In [ ]:
#df_music_pivoted = df_users.pivot(index='user_id', columns='track_id', values='rating').fillna(0)

We codify each user and song to a unique integer ID

In [123]:
user_codes, user_uniques = pd.factorize(df_users['user_id'])
track_codes, track_uniques = pd.factorize(df_users['track_id'])

We create the interaction matrix as a sparse csr matrix

In [125]:
interaction_matrix = csr_matrix((df_users_rating['rating'], (user_codes, track_codes)),shape=(len(user_uniques), len(track_uniques)))

In [126]:
# Calcular el espacio en memoria ocupado por interaction_matrix en MB
memory_usage_bytes = (
    interaction_matrix.data.nbytes +
    interaction_matrix.indptr.nbytes +
    interaction_matrix.indices.nbytes
)
memory_usage_mb = memory_usage_bytes / (1024 ** 2)
print(f"Espacio en memoria de interaction_matrix: {memory_usage_mb:.2f} MB")

Espacio en memoria de interaction_matrix: 114.81 MB


The sparse matrix has a size of only 114.81 MB, while the normal matrix as a dataframe was 218 GB

In [127]:
interaction_matrix[:10, 30:60].todense()

matrix([[0.        , 0.        , 0.        , 0.        , 0.        ,
         0.        , 0.        , 0.        , 0.        , 0.        ,
         0.        , 0.        , 0.        , 0.        , 0.        ,
         0.        , 0.        , 0.        , 0.        , 0.        ,
         0.        , 0.        , 0.        , 0.        , 0.        ,
         0.        , 0.        , 0.        , 0.        , 0.        ],
        [0.        , 0.        , 0.        , 0.        , 0.        ,
         0.        , 0.        , 0.        , 0.        , 0.        ,
         0.        , 0.        , 0.        , 0.        , 0.        ,
         0.        , 0.        , 0.        , 0.        , 0.        ,
         0.        , 0.        , 0.        , 0.        , 0.        ,
         0.        , 0.        , 0.        , 0.        , 0.        ],
        [0.        , 0.        , 0.        , 0.        , 0.        ,
         0.        , 0.        , 0.        , 0.        , 0.        ,
         0.        , 0.        ,